src/EventListener/ExceptionListener.php

```diff
@@ -0,0 +1,416 @@
+            }
+        } while (null !== ($exception = $exception->getPrevious()));
+
+        return null;
+    }
+
+    /**
+     * Check if the Contao 404 should be rendered.
+     *
+     * @param \Exception $exception The exception to test.
+     *
+     * @return bool
+     */
+    private function isRenderingOfContao404PossibleFor($exception)
+    {
+        // If not an 404 exception or the frontend has not been booted, we can not render.
+        if (!($exception instanceof NotFoundHttpException && defined('BE_USER_LOGGED_IN'))) {
+            return false;
+        }
+
+        // It is hopeless to render a 404 page when no root page or no page at all is published.
+        if ($exception instanceof RootNotFoundHttpException || $exception instanceof NoPagesFoundHttpException) {
+            return false;
+        }
+
+        // If no 404 page handler has been registered, we also can not render.
+        if (!(isset($GLOBALS['TL_PTY']['error_404']) && class_exists($GLOBALS['TL_PTY']['error_404']))) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Try to render a Contao 404 page first.
+     *
+     * @param \Exception $exception The exception to check.
+     *
+     * @return null|Response
+     *
+     * @throws \Exception When anything went really haywire during rendering of the page.
+     */
+    private function tryToRenderContao404($exception)
+    {
+        if (!$this->isRenderingOfContao404PossibleFor($exception)) {
+            return null;
+        }
+
+        static $processing;
+        // Exit when we are already trying to render a 404 page but an exception was thrown from within.
+        if ($processing) {
+            return null;
+        }
+
+        // Prevent entering this method multiple times causing endless loop.
+        $processing = true;
+
+        try {
+            // FIXME: introduce some contao_frontend_404 route and do a subrequest on it might be more sufficient.
+            /** @var \PageError404 $pageHandler */
+            $pageHandler = new $GLOBALS['TL_PTY']['error_404']();
+            $response    = $pageHandler->getResponse(false);
+            $processing  = false;
+
+            return $response;
+        } catch (ResponseException $pageException) {
+            $processing = false;
+
+            return $pageException->getResponse();
+        } catch (NotFoundHttpException $pageException) {
+            $processing = false;
+
+            // We can safely assume that when we get an not found exception when trying to render a not found
+            // page, that we have to render the internal page.
+            return null;
+        } catch (\Exception $pageException) {
+            $processing = false;
+            // Throwing the page exception here is intentional as we have some error handler problems!
+            // We should never end up here.
+            // Fixing error handling code paths is more important than the original exception.
```

Comment:
Right, but as written in the comment this indicates an bug in the exception rendering itself and therefore we do not render the pretty screen.
We might want to change this into some external handler as written in the FIXME in the future though.